File: LeetCode/117PopulatingNextRightPointersInEachNodeII/PopulatingNextRightPointersInEachNodeII.cpp
class Solution {
public:
    Node* connect(Node* root) {
        // BFS -> Time:O(n), Space:O(n)
        if (root) {
            queue<Node*> q;
            q.push(root);
            Node *front, *prev;
            while (!q.empty()) {
                int n = q.size();
                prev = NULL;
                for (int i = 0; i < n; i++) {
                    front = q.front();
                    q.pop();
                    front->next = prev;
                    prev = front;
                    if (front->right) {
                        q.push(front->right);
                    }
                    if (front->left) {
                        q.push(front->left);
                    }
                }
            }
        }
        return root;
    }
};

File: LeetCode/117PopulatingNextRightPointersInEachNodeII/PopulatingNextRightPointersInEachNodeII.py
class Solution:
    def connect(self, root: 'Node') -> 'Node':
        # BFS -> Time:O(n), Space:O(n)
        if root:
            q = []
            q.append(root)
            while q:
                prev = None
                for i in range(len(q)):
                    top = q.pop(0)
                    top.next = prev
                    prev = top
                    if top.right:
                        q.append(top.right)
                    if top.left:
                        q.append(top.left)
        return root
